### ISSUE TITLE
Add Lambda Logging Support

### ISSUE THREAD

**Issue by eahefnawy** [serverless/serverless#1654]
Add a new serverless command to show the logs of a certain function, optionally in a specific stage/region + tailing:

```
serverless logs --function --stage --region --tail
```

This command fetches the cloudwatch logs of the provided lambda function. Optionally, if the user provided the tail option, the log output will be persistent in the terminal session.

cc/ @flomotlik @pmuens

**Comment by eahefnawy** [comment]
More Options provided by AWS SDK:

- filterPattern: a pattern to filter the output of logs requested from CloudWatch
- duration: get the logs of the past hour for example
- pollInterval: control how often we check for new logs with the `--tail` option.

**Comment by MrRio** [comment]
I like the sound of this

**Comment by pmuens** [comment]
Is this logging function specific or does ist also support service level logging?

**Comment by eahefnawy** [comment]
hmmm I don't see a need for service level logging, it'd be way too noisy to combine all functions logs together in a single terminal session imo. Plus, we'll have to hack our way into supporting it since it's not really possible to combine the logs of multiple functions in AWS.

**Comment by flomotlik** [comment]
Imho Service level logging can be really helpful when you want to have a good understanding of the whole service and whats happening there. Basically tracing requests throughout the whole system.

From my perspective this is ready to be implemented. Service wide logging should be researched more to make sure how we can do this, but other than that I'd say lets get this started.

**Comment by mthenw** [comment]
> Imho Service level logging can be really helpful when you want to have a good understanding of the whole service and whats happening there. Basically tracing requests throughout the whole system. 

Same here.

Also would be great to specify starting date/time for logs:

`serverless logs -f hello --start_at <date here>`

**Comment by eahefnawy** [comment]
@mthenw you can set a `duration` with:

```
serverless logs -f hello -d 10m
```
Which will get the logs in the past 10 minutes ... it's an abstraction layer on top of `starts_at` with the help of moment.js

**Comment by flomotlik** [comment]
@eahefnawy is that duration something we implemented or that is in the SDK? Is it possible to define specific times, e.g. --start_at --end_at?

Sometimes when you want to go back 10 hours, but then only see like 15 minutes of logs thats really helpful. If its not easy then we might not want to do it, but it would be nice to have imho

**Comment by eahefnawy** [comment]
Ok so I looked a bit deeper on service wide logs, and it's definitely possible. Each Lambda function has its own logGroup with the name `/aws/lambda/lambdaName`, which we use to get the log stream in the logging plugin based on the provided function.

To get the logs of all functions we can simply iterate over all functions and get all the log groups. Pretty simple, but could be too much API requests (we can always do it Async though). Another way is to request all the log groups in an AWS account and filter them by the functions that we have. There are helpful methods in the AWS SDK for both of these approaches.

Either way, once we know the log groups we need, we'll have to request the **log streams** of each group **individually** (another Async call?), since AWS doesn't provide a patch operation for that.Then comes the hard part -> displaying all this info in the terminal.

We'll then have to think of a good way to organize all this info in the terminal screen. We can put all the logs of the same function together and separate by space or something, but either way, it'll be a lot of info to display top to bottom imo.

The tailing option also adds to this complexity, since tailing basically works by polling the log stream at a specific interval. Now that we have lots of log groups, polling all that could be really slow.

So in summary, it's definitely possible, but could be slow and the output could be just too much to handle.

**Comment by eahefnawy** [comment]
@flomotlik the SDK `filterLogEvents` method has a native `startTime` option that you pass in the params, which is just the unix epoch of the start time. The duration is an abstraction layer that lets users pass a simple `5m` or `10h` or `3d` strings in the CLI instead of passing an epoch.

I think this is much simpler to pass in the CLI than epoch or even standard date/time string that we later convert to epoch imo.

**Comment by flomotlik** [comment]
So you can do startTime and duration, which would basically be what I mentioned before. That sounds really good imho. And we also have a much easier time as we don't have to parse any dates or anything, we just tell people to use the epoch

>Another way is to request all the log groups in an AWS account and filter them by the functions that we have. There are helpful methods in the AWS SDK for both of these approaches.

Would this filtering happen in AWS or through the SDK locally? Because if it happens locally that could be a lot of work locally and probably better to only pull in specific ones.

> We'll then have to think of a good way to organize all this info in the terminal screen. We can put all the logs of the same function together and separate by space or something, but either way, it'll be a lot of info to display top to bottom imo

Can we simply sort them by date and print it? So we're pulling the same timeframe from every service and then simply print it directly?

Best case AWS can even do the combination and ordering for us.

But it sounds to me like this could be an additional feature in the future and we might want to focus on just showing the log of one function at a time first and then learn more from users once its out and improve?

**Comment by eahefnawy** [comment]
> Would this filtering happen in AWS or through the SDK locally? Because if it happens locally that could be a lot of work locally and probably better to only pull in specific ones.

This filtering would happen locally since the SDK method just returns all log groups in an AWS account, including logs in different services, random non-serverless lambdas and any logs basically.

> Can we simply sort them by date and print it? So we're pulling the same timeframe from every service and then simply print it directly?

Yep, each log stream comes with a timestamp that we can use for sorting. One thing to keep in mind though is that each lambda invocation automatically logs three default streams: `START`, `END` & `REPORT` ... so there's a lot of those in a single function and even more in a single service ... sorting by timestamp with those included could get mixed up, specially with cases where different invocations happening at the same time.

> Best case AWS can even do the combination and ordering for us.

You can order logs streams of each log group either descending or ascending, but then again the issue is using **multiple log groups**, which AWS doesn't provide natively. We're gonna have to do that on our own.

> But it sounds to me like this could be an additional feature in the future and we might want to focus on just showing the log of one function at a time first and then learn more from users once its out and improve?

Yeah imo we should just wait for now and see if people actually requests this enhancement

**Comment by flomotlik** [comment]
> Yeah imo we should just wait for now and see if people actually requests this enhancement

Then lets do this :)

**Comment by eahefnawy** [comment]
Implemented in [PR 1691](https://github.com/serverless/serverless/pull/1691)
